plugin/src/main/java/io/jenkins/plugins/redis/RedisFingerprintStorage.java

```diff
@@ -0,0 +1,168 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package io.jenkins.plugins.redis;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import hudson.Extension;
+import hudson.ExtensionList;
+import jenkins.fingerprints.FingerprintStorage;
+import hudson.model.Fingerprint;
+import hudson.Util;
+
+import java.io.IOException;
+import java.io.ByteArrayInputStream;
+import java.io.StringWriter;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.logging.Logger;
+import java.util.logging.Level;
+
+import jenkins.model.GlobalConfiguration;
+import org.jenkinsci.main.modules.instance_identity.InstanceIdentity;
+
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+import redis.clients.jedis.exceptions.JedisException;
+
+/**
+ * Pluggable external fingerprint storage for fingerprints in Redis.
+ */
+
+@Extension
+public class RedisFingerprintStorage extends FingerprintStorage {
+
+    private final String instanceId;
+    private static final Logger logger = Logger.getLogger(Fingerprint.class.getName());
+    private static volatile JedisPool jedisPool;
+
+    public static RedisFingerprintStorage get() {
+        return ExtensionList.lookup(RedisFingerprintStorage.class).get(0);
+    }
+
+    public RedisFingerprintStorage() throws IOException {
+        try {
+            instanceId = Util.getDigestOf(new ByteArrayInputStream(InstanceIdentity.get().getPublic().getEncoded()));
+            createJedisPoolFromConfig();
+        } catch (IOException e) {
+            logger.log(Level.WARNING, "Failed to obtain Instance ID. " + e);
+            throw e;
+        }
+    }
+
+    void createJedisPoolFromConfig() {
+        GlobalRedisConfiguration config = GlobalRedisConfiguration.get();
+        createJedisPool(config.getHost(), config.getPort(), config.getConnectionTimeout(), config.getSocketTimeout(),
+                config.getUsername(), config.getPassword(), config.getDatabase(), config.getSsl());
+    }
+
+    void createJedisPool(
+            String host, int port, int connectionTimeout, int socketTimeout, String username, String password,
+            int database, boolean ssl) {
+        jedisPool = new JedisPool(new JedisPoolConfig(), host, port, connectionTimeout, socketTimeout, username,
+                password, database, "Jenkins", ssl);
+    }
+
+    private @NonNull Jedis getJedis() throws JedisException{
+        return jedisPool.getResource();
+    }
+
+    /**
+     * Saves the given fingerprint.
+     */
+    public synchronized void save(Fingerprint fp) throws JedisException {
+        Jedis jedis = null;
+        StringWriter writer = new StringWriter();
+        Fingerprint.getXStream().toXML(fp, writer);
+        try {
+            jedis = getJedis();
```

Comment:
Strange that it is not autocloseable. If so, maybe it makes sense to add a utility class so that try-with-resources is used here and in other calls